src/main/java/shef/servlets/DisplayRecipesServlet.java

```diff
@@ -0,0 +1,86 @@
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import shef.data.Recipe;
+import shef.data.Step;
+import shef.servlets.NewRecipeServlet;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+ 
+/** Servlet responsible for displaying comments. */
+@WebServlet("/display-recipes")
+public class DisplayRecipesServlet extends HttpServlet {
+ 
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Recipe").addSort("timestamp", SortDirection.DESCENDING);
+ 
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+ 
+    List<Recipe> recipes = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      recipes.add(entityToRecipe(entity));
+    }
+ 
+    Gson gson = new Gson();
+ 
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(recipes));
+  }
+
+   /** Converts a Datastore entity into a Recipe. */
+  public Recipe entityToRecipe(Entity recipeEntity) {
+    String key = KeyFactory.keyToString(recipeEntity.getKey());
+    String name = (String) recipeEntity.getProperty("name");
+    String description = (String) recipeEntity.getProperty("description");
+    LinkedHashSet<String> tags = new LinkedHashSet<>((LinkedList<String>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("tags"), "tag"));
+    LinkedHashSet<String> ingredients = new LinkedHashSet<>((LinkedList<String>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("ingredients"), "ingredient"));
+    LinkedList<Step> steps = (LinkedList<Step>) (LinkedList<?>) getDataAsList(recipeEntity.getProperty("steps"), "step");
+    long timestamp = (long) recipeEntity.getProperty("timestamp");
+    return new Recipe(key, name, description, tags, ingredients, steps, timestamp);
+  }
+
+  /** Gets a list of Recipe parameters from a Datastore property. */
+  public Collection<Object> getDataAsList(Object propertiesObject, String field) {
+    Collection<EmbeddedEntity> properties = (Collection<EmbeddedEntity>) propertiesObject;
+    Collection<Object> dataAsList = new LinkedList<>();
```

Comment:
The method does so to be generalized and applicable to tags, ingredients, and steps. Tags and ingredients use HashSet, but steps use LinkedList, so the method returns a Collection to satisfy them all. Additionally, tags and ingredients are strings, but steps have their own object, so the method returns a collection of objects. This is an old messy method, and PR #81 introduces a Recipe constructor from a Datastore entity that does this all cleanly.

Reply:
^Just to add onto this - we can either get that PR in first and then I'll make the changes downstream to this, too, or we can just keep this as is so that we can deploy the MVP, and then after #81 is in, I'll create another PR that fixes this.